File: client_lib/java/src/com/basho/riak/JiakClient.java
/*
This file is provided to you under the Apache License,
Version 2.0 (the "License"); you may not use this file
except in compliance with the License.  You may obtain
a copy of the License at
   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.  
 */
package com.basho.riak;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;

import sun.net.www.protocol.http.HttpURLConnection;

/**
 * Client class for accessing the Riak document store via the HTTP/JSON
 * interface.
 * 
 * @author Andy Gross <andy@basho.com>
 * @version 0.1
 */
public class JiakClient {

	private final String ip;
	private final String port;
	private final String prefix;

	/**
	 * Construct a JiakClient with the default server path prefix of /jiak/
	 * 
	 * @param ip
	 *            The IP address of the Riak server
	 * @param port
	 *            The Jiak web port to access
	 */
	public JiakClient(final String ip, final String port) {
		this(ip, port, "/jiak/");
	}

	/**
	 * Construct a JiakClient with a user-specified path prefix
	 * 
	 * @param ip
	 *            The IP address of the Riak server.
	 * @param port
	 *            The Jiak web port to access.
	 * @param prefix
	 *            The server path prefix.
	 */
	public JiakClient(final String ip, final String port, final String prefix) {
		this.ip = ip;
		this.port = port;
		this.prefix = prefix;
	}

	/**
	 * Set the schema describing the structure and per-field permissions for a
	 * Riak bucket.
	 * 
	 * @param bucket
	 *            The bucket name.
	 * @param allowedFields
	 *            The complete list of field names allowed in a document.
	 * @param writeMask
	 *            The list of writable fields, or null. If null, the write mask
	 *            is set to the value of <code>allowedFields</code>.
	 * @param readMask
	 *            The list of readable fields, or null. If null, the read mask
	 *            is set to the value of <code>allowedFields</code>.
	 * @param requiredFields
	 *            The list of fields that must be present when creating or
	 *            updating a document. If null, <code>requiredFields</code> is
	 *            set to an empty list.
	 * 
	 * @throws JSONException
	 *             If an error occurs assembling the JSON request body.
	 * @throws IOException
	 *             If an error occurs during communication with the Riak server.
	 * @throws JiakException
	 *             If the Riak server returns an error or or an unexpected
	 *             response code.
	 */
	public void setBucketSchema(final String bucket,
			final List<String> allowedFields, List<String> writeMask,
			List<String> readMask, List<String> requiredFields)
			throws JSONException, IOException, JiakException {

		if (requiredFields == null)
			requiredFields = new ArrayList<String>();
		if (writeMask == null)
			writeMask = new ArrayList<String>(allowedFields);
		if (readMask == null)
			readMask = new ArrayList<String>(allowedFields);

		final JSONObject schema = new JSONObject();
		schema.put("allowed_fields", allowedFields);
		schema.put("required_fields", requiredFields);
		schema.put("read_mask", readMask);
		schema.put("write_mask", writeMask);
		final JSONObject schemaReqBody = new JSONObject();
		schemaReqBody.put("schema", schema);
		final HashMap<String, String> reqHeaders = new HashMap<String, String>();
		reqHeaders.put("Content-Type", "application/json");
		reqHeaders.put("Accept", "application/json");
		final String reqURI = makeURI(bucket);
		final HttpURLConnection requestConn = doRequest("PUT", reqURI,
				schemaReqBody, reqHeaders);
		if (requestConn.getResponseCode() != 204)
			throw new JiakException(requestConn.getResponseMessage());
	}

	/**
	 * Return the schema and keys for a Riak bucket.
	 * 
	 * @param bucket
	 *            The bucket to list.
	 * @return A <code>JSONObject</code> with keys <code>schema</code> and
	 *         <code>keys</code>.
	 * @throws JSONException
	 *             If an error occurs assembling the JSON request body.
	 * @throws IOException
	 *             If an error occurs during communication with the Riak server.
	 * @throws JiakException
	 *             If the Riak server returns an error or or an unexpected
	 *             response code.
	 */
	public JSONObject listBucket(final String bucket) throws IOException,
			JSONException, JiakException {
		final HashMap<String, String> reqHeaders = new HashMap<String, String>();
		reqHeaders.put("Content-Type", "application/json");
		reqHeaders.put("Accept", "application/json");
		final HttpURLConnection requestConn = doRequest("GET", makeURI(bucket),
				null, reqHeaders);
		return expect(200, requestConn);
	}

	/**
	 * Store a {@link JiakObject}.
	 * 
	 * @param object
	 *            The {@link JiakObject} to store.
	 * 
	 * @throws JSONException
	 *             If an error occurs assembling the JSON request body.
	 * @throws IOException
	 *             If an error occurs during communication with the Riak server.
	 * @throws JiakException
	 *             If the Riak server returns an error or or an unexpected
	 *             response code.
	 */
	public void store(final JiakObject object) throws IOException,
			JSONException, JiakException {
		final HashMap<String, String> reqHeaders = new HashMap<String, String>();
		reqHeaders.put("Content-Type", "application/json");
		reqHeaders.put("Accept", "application/json");
		final String reqURI = makeURI(object.getBucket() + "/"
				+ object.getKey() + "?returnbody=true");
		final HttpURLConnection requestConn = doRequest("PUT", reqURI, object
				.toJSONObject(), reqHeaders);
		final JSONObject updated = expect(200, requestConn);
		object.update(updated);
	}

	/**
	 * Fetch the {@link JiakObject} stored at <code>bucket</code> and
	 * <code>key</code>.
	 * 
	 * @param bucket
	 *            The bucket containing the {@link JiakObject} to fetch.
	 * @param key
	 *            The key of the {@link JiakObject} to fetch.
	 * 
	 * @return A {@link JiakObject}.
	 * 
	 * @throws JSONException
	 *             If an error occurs assembling the JSON request body.
	 * @throws IOException
	 *             If an error occurs during communication with the Riak server.
	 * @throws JiakException
	 *             If the Riak server returns an error or or an unexpected
	 *             response code.
	 */
	public JiakObject fetch(final String bucket, final String key)
			throws IOException, JSONException, JiakException {
		final HashMap<String, String> reqHeaders = new HashMap<String, String>();
		reqHeaders.put("Content-Type", "application/json");
		reqHeaders.put("Accept", "application/json");
		final String reqURI = makeURI(bucket + "/" + key);
		final HttpURLConnection requestConn = doRequest("GET", reqURI, null,
				reqHeaders);
		if (requestConn.getResponseCode() == 404)
			return null;
		final JSONObject objData = expect(200, requestConn);
		final JiakObject object = new JiakObject(bucket, key);
		object.update(objData);
		return object;
	}

	/**
	 * Delete a {@link JiakObject}.
	 * 
	 * @param bucket
	 *            The bucket containing the object to delete.
	 * @param key
	 *            The key of the object to delete.
	 * 
	 * @throws JSONException
	 *             If an error occurs assembling the JSON request body.
	 * @throws IOException
	 *             If an error occurs during communication with the Riak server.
	 * @throws JiakException
	 *             If the Riak server returns an error or or an unexpected
	 *             response code.
	 */
	public void delete(final String bucket, final String key)
			throws IOException, JiakException {
		final String reqURI = makeURI(bucket + "/" + key);
		final HashMap<String, String> reqHeaders = new HashMap<String, String>();
		reqHeaders.put("Accept", "*/*");
		final HttpURLConnection requestConn = doRequest("DELETE", reqURI, null,
				reqHeaders);
		final int responseCode = requestConn.getResponseCode();
		if (responseCode != 404 && responseCode != 204)
			throw new JiakException(requestConn.getResponseMessage());
	}

	protected HttpURLConnection doRequest(final String method,
			final String uri, final JSONObject body,
			final Map<String, String> headers) throws IOException {
		final URL requestURL = new URL(uri);
		final HttpURLConnection requestConn = (HttpURLConnection) requestURL
				.openConnection();
		requestConn.setRequestMethod(method);
		if (headers != null)
			for (final String k : headers.keySet())
				requestConn.setRequestProperty(k, headers.get(k));
		if (body != null) {
			requestConn.setDoOutput(true);
			writeRequestBody(requestConn, body);
		}
		return requestConn;
	}

	protected JSONObject expect(final int responseCode,
			final HttpURLConnection connection) throws JSONException,
			JiakException, IOException {
		if (connection.getResponseCode() == responseCode) {
			final String responseBody = readResponseBody(connection);
			return new JSONObject(responseBody);
		}
		throw new JiakException(connection.getResponseMessage());
	}

	protected String makeURI(final String path) {
		return "http://" + ip + ":" + port + prefix + path;
	}

	protected static void writeRequestBody(final URLConnection connection,
			final JSONObject body) {
		OutputStream outputStream;
		try {
			outputStream = connection.getOutputStream();
		} catch (final IOException e1) {
			e1.printStackTrace();
			return;
		}
		final OutputStreamWriter outputStreamWriter = new OutputStreamWriter(
				outputStream);
		try {
			outputStreamWriter.write(body.toString());
			outputStreamWriter.flush();
		} catch (final IOException e) {
			e.printStackTrace();
		} finally {
			try {
				outputStream.close();
			} catch (final IOException e) {
				e.printStackTrace();
			}
		}
	}

	protected static String readResponseBody(final HttpURLConnection connection) {
		String urlContent = null;
		InputStream inputStream;
		try {
			inputStream = connection.getInputStream();
		} catch (final IOException e) {
			e.printStackTrace();
			return urlContent;
		}
		try {
			urlContent = downloadStream(inputStream);
		} catch (final IOException e) {
			e.printStackTrace();
		} finally {
			try {
				inputStream.close();
			} catch (final IOException e) {
				e.printStackTrace();
			}
		}
		return urlContent;
	}

	protected static String downloadStream(final InputStream in)
			throws IOException {

		final OutputStream out = new ByteArrayOutputStream();
		try {
			copy(in, out);
		} finally {
			in.close();
			out.close();
		}
		return out.toString();

	}

	protected static void copy(final InputStream in, final OutputStream out)
			throws IOException {

		final byte[] buffer = new byte[1024];
		while (true) {
			final int readCount = in.read(buffer);
			if (readCount == -1)
				break;
			out.write(buffer, 0, readCount);
		}
	}
}

File: apps/riak/src/jiak_client.erl
%% This file is provided to you under the Apache License,
%% Version 2.0 (the "License"); you may not use this file
%% except in compliance with the License.  You may obtain
%% a copy of the License at

%%   http://www.apache.org/licenses/LICENSE-2.0

%% Unless required by applicable law or agreed to in writing,
%% software distributed under the License is distributed on an
%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
%% KIND, either express or implied.  See the License for the
%% specific language governing permissions and limitations
%% under the License.    

%% @doc The equivalent of {@link riak_client}, but for the "jiak"
%%      pattern of access.  Functions defined here are equivalent
%%      to the like-named functions in riak_client, but expect
%%      jiak-style paramenters (like binary object keys) and
%%      return {@link jiak_object}s.
%% @type jiak_client() = term()
-module(jiak_client, [RiakClient]).
-author('Bryan Fink <bryan@basho.com>').

-export([mapred/2,mapred/3]).
-export([get/3,get/4]).
-export([put/2,put/3,put/4]).
-export([delete/3,delete/4]).
-export([list_keys/1]).
-export([set_bucket/2,get_bucket/1]).
-export([reload_all/1]).
-export([riak_client/0]).

%% @type default_timeout() = 15000
-define(DEFAULT_TIMEOUT, 15000).

%% @see riak_object:mapred/2
%% @equiv mapred(Inputs, Query, default_timeout())
mapred(Inputs,Query) -> mapred(Inputs, Query, ?DEFAULT_TIMEOUT).

%% @see riak_object:mapred/3
%% @equiv riak_client:mapred(Inputs, Query, Timeout)
mapred(Inputs,Query,Timeout) ->
    RiakClient:mapred(Inputs,Query,Timeout).

%% @spec get(riak_object:bucket(), jiak_object:key(), R :: integer()) ->
%%       {ok, jiak_object()} |
%%       {error, notfound} |
%%       {error, timeout} |
%%       {error, Err :: term()}
%% @doc Fetch the object at Bucket/Key.  Return a value as soon as R
%%      nodes have responded with a value or error.
%% @equiv get(Bucket, Key, R, default_timeout())
get(Bucket, Key, R) -> get(Bucket, Key, R, ?DEFAULT_TIMEOUT).

%% @spec get(riak_object:riak_object(), jiak_object:key(), R :: integer(),
%%           TimeoutMillisecs :: integer()) ->
%%       {ok, jiak_object()} |
%%       {error, notfound} |
%%       {error, timeout} |
%%       {error, Err :: term()}
%% @doc Fetch the object at Bucket/Key.  Return a value as soon as R
%%      nodes have responded with a value or error, or TimeoutMillisecs passes.
get(Bucket, Key, R, Timeout) when is_binary(Bucket), is_binary(Key) ->
    case RiakClient:get(Bucket, Key, R, Timeout) of
        {ok, RiakObject} ->
            JiakObject = jiak_object:from_riak_object(RiakObject),
            {ok, JiakObject};
        Error ->
            Error
    end.

%% @spec put(JiakObject :: jiak_object(), W :: integer()) ->
%%        ok |
%%       {error, too_many_fails} |
%%       {error, timeout}
%% @doc Store RObj in the cluster.
%%      Return as soon as at least W nodes have received the request.
%% @equiv put(RObj, W, W, default_timeout())
put(JiakObject, W) -> put(JiakObject, W, W, ?DEFAULT_TIMEOUT).

%% @spec put(JiakObject :: jiak_object(), W :: integer(), RW :: integer()) ->
%%        ok |
%%       {error, too_many_fails} |
%%       {error, timeout}
%% @doc Store RObj in the cluster.
%%      Return as soon as at least W nodes have received the request, and
%%      at least DW nodes have stored it in their storage backend.
%% @equiv put(Robj, W, DW, default_timeout())
put(JiakObject, W, DW) -> put(JiakObject, W, DW, ?DEFAULT_TIMEOUT).

%% @spec put(JiakObject :: jiak_object(), W :: integer(), RW :: integer(),
%%           TimeoutMillisecs :: integer()) ->
%%        ok |
%%       {error, too_many_fails} |
%%       {error, timeout}
%% @doc Store RObj in the cluster.
%%      Return as soon as at least W nodes have received the request, and
%%      at least DW nodes have stored it in their storage backend, or
%%      TimeoutMillisecs passes.
put(JiakObject, W, DW, Timeout) ->    
    RiakObject = jiak_object:to_riak_object(JiakObject),
    RiakClient:put(RiakObject, W, DW, Timeout).

%% @spec delete(riak_object:bucket(), jiak_object:key(), RW :: integer()) ->
%%        ok |
%%       {error, too_many_fails} |
%%       {error, notfound} |
%%       {error, timeout} |
%%       {error, Err :: term()}
%% @doc Delete the object at Bucket/Key.  Return a value as soon as RW
%%      nodes have responded with a value or error.
%% @equiv delete(Bucket, Key, RW, default_timeout())
delete(Bucket,Key,RW) -> delete(Bucket,Key,RW,?DEFAULT_TIMEOUT).

%% @spec delete(niak_object:bucket(), jiak_object:key(), RW :: integer(),
%%           TimeoutMillisecs :: integer()) ->
%%        ok |
%%       {error, too_many_fails} |
%%       {error, notfound} |
%%       {error, timeout} |
%%       {error, Err :: term()}
%% @doc Delete the object at Bucket/Key.  Return a value as soon as RW
%%      nodes have responded with a value or error, or TimeoutMillisecs passes.
%% @equiv riak_client:delete(Bucket, Key, RW, Timeout)
delete(Bucket,Key,RW,Timeout) ->
    RiakClient:delete(Bucket, Key, RW, Timeout).

%% @spec list_keys(riak_object:bucket()) ->
%%       {ok, [Key :: riak_object:key()]} |
%%       {error, timeout} |
%%       {error, Err :: term()}
%% @doc List the keys known to be present in Bucket.
%%      Key lists are updated asynchronously, so this may be slightly
%%      out of date if called immediately after a put or delete.
%% @equiv riak_client:list_keys(Bucket)
list_keys(Bucket) -> RiakClient:list_keys(Bucket).

%% @spec set_bucket(riak_object:bucket(), [BucketProp :: {atom(),term()}]) -> ok
%% @doc Set the given properties for Bucket.
%% @equiv riak_client:set_bucket(BucketName, BucketProps)
set_bucket(BucketName,BucketProps) ->
    RiakClient:set_bucket(BucketName, BucketProps).

%% @spec get_bucket(riak_object:bucket()) -> [BucketProp :: {atom(),term()}]
%% @doc Get all properties for Bucket.
%% @equiv riak_client:get_bucket(BucketName)
get_bucket(BucketName) -> RiakClient:get_bucket(BucketName).

%% @spec reload_all(Module :: atom()) -> term()
%% @doc Force all Riak nodes to reload Module.
%%      This is used when loading new modules for map/reduce functionality.
%% @equiv riak_client:reload_all(Module)
reload_all(Module) -> RiakClient:reload_all(Module).

%% @spec riak_client() -> riak_client()
%% @doc Get the riak_client that this jiak_client is using.
riak_client() -> RiakClient.
